cmake/modules/extensions.cmake

```diff
@@ -109,6 +109,7 @@ endfunction()
 
 # https://cmake.org/cmake/help/latest/command/target_link_libraries.html
 function(zephyr_link_libraries)
+  set_property(TARGET zephyr_interface APPEND PROPERTY LINK_LIBS ${ARGV})
```

Comment:
Why do we need to maintain this property ?

Secondly, this will probably break when using generator expressions with `zephyr_link_libraries()` as generator expressions cannot be correctly retrieved on custom properties, only some properties support genex.

If this is the reason:
> As many libraries used in zephyr_link_libraries depend upon C library
and/or kernel interfaces, they need to be present in the link command line
before those libaries.

then what should be done is this:
```
# This is a regular lib depending on Kernel
add_library(some_lib) 

target_link_library(some_lib PUBLIC kernel)
```

that will ensure the following linking order `... some_lib kernel ...` which seems to be what you want to achieve.

@keith-packard can you provide the an example of what is breaking today, cause this proposal could indicate that it's as simple as missing such link dependency.

And I will be happy to assist.

This is one downside of the `zephyr_library()`, that users never get to understand the internals of CMake's linking order because things are whole-archived together, and when they suddenly are not, things break.

Reply:
As above, there are two problems with trying to use direct dependencies. The first is toolchain libraries, like `-lstdc++` and `-lc` where we cannot use add_library without kludges. The second is that every attempt I made to do this resulted in cmake finding a circular dependency that included `kernel`, `offsets_h`, `offsets`, and `zephyr_interface`.